### PR TITLE
RigRetriever class

### PR DESCRIPTION
**Pull request by dellaert**
A retriever class that reads a file of constraints, including a set of 5 x 5 counts, indicating which cameras on the rig in two successive poses are likely to have visual matches. These counts are matching features from the lidar, in the case of the Hilti challenge, and serve us a proxy to see how good we expect visual matching to work between two cameras.

Currently hardcoded to 5 cameras. There is no need to generalize yet.

### CODE REVIEW

**Comment by dellaert** [comment]
@womackj1 FYI
@jerredchen I updated your code and unit tests. I suggest you copy back to the SLAM repo.

**Comment by dellaert** [comment]
PS @ayushbaid how do I check in a new data directory? Tee tests will fail as somehow the files I added in `tests/data/hilti` are not seen as added files by git.

**Comment by ayushbaid** [comment]
> PS @ayushbaid how do I check in a new data directory? Tee tests will fail as somehow the files I added in `tests/data/hilti` are not seen as added files by git.

I use the `git add -f` to add a small subset of the test directory.

**Comment by ayushbaid** [comment]
@dellaert I assume the final version of this PR will be on the Hilti loader (which I will push once I we fix the time sync issue)?

File: gtsfm/retriever/rig_retriever.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+from gtsfm.retriever.retriever_base import RetrieverBase
+
+
+@dataclass
+class Constraint:
+    a: int = 0
+    b: int = 1
+    aTb: gtsam.Pose3 = gtsam.Pose3()
+    cov: np.ndarray = np.eye(6)
+    counts: np.ndarray = np.zeros((5, 5))
```
Comment (ayushbaid) [inline gtsfm/retriever/rig_retriever.py:22]: Can you add a docstring for the args. `counts` for example is non-trivial to understand.

File: gtsfm/retriever/rig_retriever.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import gtsam
+import numpy as np
+from gtsfm.loader.loader_base import LoaderBase
+from gtsfm.retriever.retriever_base import RetrieverBase
+
+
+@dataclass
+class Constraint:
+    a: int = 0
+    b: int = 1
```
Comment (ayushbaid) [inline gtsfm/retriever/rig_retriever.py:19]: Should we skip default values for a and b?

File: gtsfm/retriever/rig_retriever.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    @staticmethod
+    def read(fname: str) -> List["Constraint"]:
+        """Read constraints from text file."""
+        constraint_matrix = np.loadtxt(fname)
+        return [Constraint.from_row(row) for row in constraint_matrix]
+
+    def edges(self, threshold=30):
+        """Return pairs of cameras with over `threshold` count."""
+        return [
+            (self.a * 5 + camera_index_in_a, self.b * 5 + camera_index_in_b)
```
Comment (ayushbaid) [inline gtsfm/retriever/rig_retriever.py:72]: Nit: moving 5 to a global variable.

File: gtsfm/retriever/rig_retriever.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    def __init__(self, constraints_path: Path, threshold: int = 100):
+        """Initialize with path to a constraints file."""
+        self._constraints_path = constraints_path
+        self._threshold = threshold
+
+    def run(self, loader: LoaderBase) -> List[Tuple[int, int]]:
+        """Compute potential image pairs.
+
+        Args:
+            loader: image loader.
```
Comment (ayushbaid) [inline gtsfm/retriever/rig_retriever.py:91]: Loader is not being used right now?

File: gtsfm/retriever/rig_retriever.py (hunk 2 above)
Comment (dellaert) [reply gtsfm/retriever/rig_retriever.py:19]: done

File: gtsfm/retriever/rig_retriever.py (hunk 1 above)
Comment (dellaert) [reply gtsfm/retriever/rig_retriever.py:22]: Done

File: gtsfm/retriever/rig_retriever.py (hunk 3 above)
Comment (dellaert) [reply gtsfm/retriever/rig_retriever.py:72]: Hardcoding 5 until we do another rig. Building for the present.

File: gtsfm/retriever/rig_retriever.py (hunk 4 above)
Comment (dellaert) [reply gtsfm/retriever/rig_retriever.py:91]: Indeed. But needed by base class

**Comment by dellaert** [comment]
@ayushbaid I'd like to integrate the real HiltiLoader in a separate PR. I think we need to make changes to it to read the constraints in this PR, as well.

**Comment by ayushbaid** [comment]
> @ayushbaid I'd like to integrate the real HiltiLoader in a separate PR. I think we need to make changes to it to read the constraints in that PR, as well.

Constraints will be read in the retriever class, right?

**Review by ayushbaid — APPROVED** [review]
LGTM

**Pull request merged**